### PR TITLE
Convert exception throwing JIT helpers to managed code.

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/ThrowHelper.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/ThrowHelper.cs:4]: This list of usings looks too long.

File: src/coreclr/vm/ecall.cpp [hunk 2]
```diff
@@ -186,6 +186,54 @@ void ECall::PopulateManagedHelpers()
     pMD = CoreLibBinder::GetMethod((BinderMethodID)(METHOD__MATH__CONVERT_TO_UINT64_CHECKED));
     pDest = pMD->GetMultiCallableAddrOfCode();
     SetJitHelperFunction(CORINFO_HELP_DBL2ULNG_OVF, pDest);
+
+    pMD = CoreLibBinder::GetMethod((BinderMethodID)(METHOD__THROW_HELPERS__THROWARGUMENTEXCEPTION));
```
Comment (jkotas) [inline src/coreclr/vm/ecall.cpp:190]: To your earlier question about this being executed during startup for no good reason, it should be a straightforward change to do this lazily only once somebody asks for the address of the JIT helper (typical app will never ask for address for most of these throwing helpers):

- There are the 3 places in jitinterface.cpp that needs to be updated to call the lazy initialization:

            Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);

        pfnHelper = hlpDynamicFuncTable[dynamicFtnNum].pfnHelper;

        pfnHelper = hlpDynamicFuncTable[((size_t)pfnHelper - 1)].pfnHelper;

- The lazy initialization would convert CorInfoHelpFunc to BinderMethodID, get the helper entrypoint (`CoreLibBinder::GetMethod`, `GetMultiCallableAddrOfCode`) and cache the pointer in `hlpDynamicFuncTable`

File: src/coreclr/vm/methodtable.cpp [hunk 3]
```diff
@@ -5188,7 +5188,7 @@ BOOL MethodTable::FindDispatchEntry(UINT32 typeID,
 
 #ifndef DACCESS_COMPILE
 
-void ThrowExceptionForAbstractOverride(
+void ThrowEntryPointNotFoundException(
```
Comment (jkotas) [inline src/coreclr/vm/methodtable.cpp:5191]: The contract for this method is wrong. It throws and allocates, so it cannot be LIMITED_METHOD_CONTRACT.

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/methodtable.cpp:5191]: Yep. I've a bit of clean-up to do in this PR before it is ready.

File: src/coreclr/System.Private.CoreLib/src/System/ThrowHelper.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/ThrowHelper.cs:4]: Yep. This was copy/paste for now.

File: src/coreclr/vm/methodtable.cpp [hunk 4]
```diff
@@ -5193,7 +5193,13 @@ void ThrowEntryPointNotFoundException(
...
     MethodDesc *pInterfaceMD)
 {
-    LIMITED_METHOD_CONTRACT;
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
```
Comment (jkotas) [inline src/coreclr/vm/methodtable.cpp:5202]:
```suggestion
    STANDARD_VM_CONTRACT;
```

File: src/coreclr/vm/methodtable.cpp [hunk 5]
```diff
@@ -5400,7 +5406,13 @@ void ThrowAmbiguousResolutionException(
...
```
Comment (jkotas) [inline src/coreclr/vm/methodtable.cpp:5415]:
```suggestion
    STANDARD_VM_CONTRACT;
```

File: src/coreclr/vm/ecall.cpp (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/ecall.cpp]: Done. This increased the PR size a bit, but it does seem to remove some cruft and reduce any concern I had with start-up impact.

File: src/coreclr/vm/jithelpers.cpp [hunk 6]
```diff
@@ -5256,28 +5027,55 @@ VMHELPDEF hlpDynamicFuncTable[DYNAMIC_CORINFO_HELP_COUNT] =
...
     _ASSERTE(ftnNum < DYNAMIC_CORINFO_HELP_COUNT);
 
     LOG((LF_JIT, LL_INFO1000000, "Setting JIT dynamic helper %3d (%s) to %p\n",
         ftnNum, hlpDynamicFuncTable[ftnNum].name, pFunc));
 
-    hlpDynamicFuncTable[ftnNum].pfnHelper = (void *) pFunc;
+    hlpDynamicFuncTable[ftnNum].pfnHelper = (void*)pFunc;
+}
+
+VMHELPDEF LoadDynamicJitHelper(DynamicCorInfoHelpFunc ftnNum)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/jithelpers.cpp:5056]: This is the main lazy load logic that was added.

**Comment by AaronRobinsonMSFT** [comment]
This is going to fail for R2R images during `Module::RunEagerFixups()` - see below. I wonder if this can be narrowed to be "valid" if the image is SPCL?

https://github.com/dotnet/runtime/blob/e95eaf06bdc2d5faeb87b586c57d301fb7f0be43/src/coreclr/vm/ceeload.cpp#L3470-L3487

File: src/coreclr/vm/jithelpers.cpp [hunk 7]
```diff
@@ -5238,16 +5001,24 @@ enum __CorInfoHelpFunc {
...
 {
-#define JITHELPER(code, pfnHelper, sig)
-#define DYNAMICJITHELPER(code, pfnHelper, sig) HELPERDEF(DYNAMIC_ ## code, pfnHelper, sig)
+#define JITHELPER(code, pfnHelper, binderId)
+#define DYNAMICJITHELPER(code, pfnHelper, binderId) HELPERDEF(DYNAMIC_ ## code, pfnHelper, binderId)
+#include "jithelpers.h"
+};
+
+// dynamic helpers to Binder ID mapping - See definition of DynamicCorInfoHelpFunc.
+static BinderMethodID hlpDynamicToBinderMap[DYNAMIC_CORINFO_HELP_COUNT] =
```
Comment (jkotas) [inline src/coreclr/vm/jithelpers.cpp:5018]:
```suggestion
static const BinderMethodID hlpDynamicToBinderMap[DYNAMIC_CORINFO_HELP_COUNT] =
```

File: src/coreclr/vm/jitinterface.cpp [hunk 8]
```diff
@@ -10550,8 +10550,8 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
                                void **            ppIndirection)  /* OUT */
 {
     CONTRACTL {
-        NOTHROW;
-        GC_NOTRIGGER;
+        THROWS;
+        GC_TRIGGERS;
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:10554]: This needs to be `JIT_TO_EE_TRANSITION` now (without _LEAF).

**Review by jkotas — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**